x/cosmwasmpool/README.md

```diff
@@ -0,0 +1,284 @@
+# CosmWasm Pool
+
+The CosmWasm Pool Module is an extension for the Osmosis pools, aiming to create a custom module that allows users to create and manage liquidity pools backed by CosmWasm smart contracts. The feature enables developers to build and deploy custom smart contracts that can be integrated with the rest of the pool types on the Osmosis chain.
+
+The module is built on top of the CosmWasm smart contracting platform, which provides a secure and efficient way to develop and execute WebAssembly (Wasm) smart contracts on the Cosmos SDK.
+
+Having pools in CosmWasm provides several benefits, one of which is avoiding the need for chain upgrades when introducing new functionalities or modifying existing ones related to liquidity pools. This advantage is particularly important in the context of speed of development and iteration.
+
+An example of a CosmWasm pool type: [transmuter](https://github.com/osmosis-labs/transmuter)
+
+
+## Creating new CosmWasm Pool
+
+To create new CosmWasm Pool, there are 3 modules involved: `x/cosmwasmpool`, `x/wasm`, and `x/poolmanager`. Here is an overview of the process:
+
+```mermaid
+graph TD;
+  Sender((Sender))
+
+  Sender -- create poool --> x/cosmwasmpool
+  x/cosmwasmpool -- get next & set pool id --> x/poolmanager
+  x/cosmwasmpool -- instantiate contract --> x/wasm
+```
+
+The CosmWasm contract that is to be instanitiated needs to implement [CosmWasm Pool Contract Interface](#cosmwasm-pool-contract-interface) and store it on chain first. Then new pool can be created by sending `MsgCreateCosmWasmPool`.
+
+
+`MsgCreateCosmWasmPool` contains `InstantiateMsg`, which is a message that will be passed to the CosmWasm contract when it is instantiated. The structure of the message is defined by the contract developer, and can contain any information that the contract needs to be instantiated. JSON format is used for `InstantiateMsg`.
```

Comment:
Should we decide on how code id should be handled?

I think we should be saving the code id on chain but having the flexibility to upgrade it via gov prop. WDYT?

Reply:
Not sure if we need to save code id if we already save address, since we can query it from wasm module unless we have explicit need to store.

For gov prop, I think it would be nice if we can have prop type where it does
- store and init
- store and migrate

Else we need 2 separated gov prop every time new pool type is added / upgrade. It might not be very frequent but if it's not too hard to do I think it make sense.

Reply:
I was thinking that it might be useful from a security standpoint.

However, on more thought, I agree that it is not needed.

We can make only admin (module account) be allowed to migrate a contract. The migration is controlled by governance.

> store code and migrate

I agree on having this.

> store code and init pool

Why would we also init pool? Shouldn't that be done by `MsgCreateCosmWasmPool` separately?

Reply:
Once we agree on this, could you please update the spec for migration logic and gov props?

Reply:
> Why would we also init pool? Shouldn't that be done by MsgCreateCosmWasmPool separately?

`MsgCreateCosmWasmPool` only create pool out of existing code stored on chain.